design/src/rfcs/rfc0026_custom_validation_exceptions.md

```diff
@@ -0,0 +1,671 @@
+mapping for concrete operations:
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .operation1(operation_handler1)
+    .operation2(operation_handler2)
+    ...
+    .operationI(operation_handlerI)
+    .on_operationI_validation_failure(map_operationI_validation_failure_to_error)
+    ...
+    .operationN(operation_handlerN)
+    .on_validation_failure(common_map_validation_failure_to_error)
+    .build()
+    .unwrap();
+```
+
+is likely to lead to confusion.
+
+#### Registering the custom validation error mapping together with the operation
+
+The mapping from an operation's `ConstraintViolationExceptions` to the
+operation error shape is likely to live next to the operation handler
+implementation. Likewise, registering the mapping and the operation handler on
+the service builder is likely to occur in neighboring lines. Since the builder
+will fail if either of the two are not provided, we could couple the
+registration of both in a single call by passing in a pair:
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .my_operation((operation_handler, map_operation_validation_failure_to_error))
+    .build()
+    .unwrap();
+```
+
+This API [resembles more][smithy-typescript-custom-validation]
+smithy-typescript's:
+
+```typescript
+const handler = getGetForecastHandler(async (input) => getForecast(input),
+    (ctx, failures) => {
+        return new BadInputError(`${failures.length} bad inputs detected.`);
+    };
+});
+```
+
+#### Synchronicity of the custom validation error mapping
+
+Should the `ConstraintViolationExceptions` mapping be `async`? In other words,
+should we require a [`core::ops::Fn`], or a future type? All use cases so far
+would be solved with a synchronous mapping, but a priori there should be no
+problem in allowing async functions. As a data point, axum's
+[`IntoResponse`](https://docs.rs/axum-core/latest/axum_core/response/trait.IntoResponse.html)
+is synchronous.
+
+#### Access to extractors in the custom validation error mapping
+
+Should the `ConstraintViolationExceptions` mapping have access to request
+extractors, like operation handlers do?
+
+```rust
+fn map_operation_validation_failure_to_error(
+    errors: input::my_operation_input::ConstraintViolationExceptions,
+    request_id: Extension<ServerRequestId>,
+    global_state: Extension<GlobalState>,
+) -> error::OperationError {
+    // ...
+}
+```
+
+Most mappings will probably be stateless and only require knowledge of the
+modeled data, but a priori there should be no problem in allowing access to
+extractors. As a data point,
+[smithy-typescript][smithy-typescript-custom-validation] does allow access to
+the incoming request's context.
+
+Since granting access to extractors is a backwards-compatible change, we could
+punt on implementing this until user interest is expressed and we learn about
+use cases.
+
+#### "Tightness" of constraint violations
+
+`ConstraintViolationExceptions` [is not
```

Comment:
What type of manipulation do we expect the users of custom validation exceptions to make that would require precise modelling of the constraint violation exception?

Reply:
In relation to https://github.com/awslabs/smithy-rs/pull/2040 - does this section advocate to use a tight exception model even if the user is not relying on custom validation exceptions?
I.e. is this why you were suggesting that this RFC is tied to https://github.com/awslabs/smithy-rs/pull/2040?

Reply:
Yes to both questions.

It's probably better that I move this tightness problem in #2040, I just thought that the problem was more apparent here.

Reply:
> What type of manipulation do we expect the users of custom validation exceptions to make that would require tight modelling of the constraint violation exception?

I don't know. It's not that they'd _require_ it. For what it's worth, an API that gave them a map from Smithy path to error message string would probably satisfy all of the use cases I know of.

Succinctly phrased, I think this is just a simple problem that raises a tradeoff between type safety/expressiveness and API complexity/understandability. But I'm not sure what option is optimal / strikes the right balance in that spectrum.

Reply:
No advantage in that case. But as the section calls out towards the end, this problem afflicts regular `ConstraintViolation`s too. I think we should only consider two cases: we either make both `ConstraintViolation`s and `ConstraintViolationException`s tight, or we make both not tight.